### PR TITLE
Subsample cluster and metadata files

### PR DESCRIPTION
**Pull request by knapii-developments**
This PR relates to [SCP-1748](https://broadworkbench.atlassian.net/browse/SCP-1748) and implements subsampling at 1K, 10K, 20K, and 100K for metadata and cluster files. There are no validation checks included and it's assumed that the cluster and metadata file are valid.

Command to run: 
python ingest_pipeline.py ingest_subsample --cluster-file ../tests/data/test_1k_cluster_Data.csv --cell-metadata-file ../tests/data/test_1k_metadata_Data.csv --subsample True

### CODE REVIEW

File: ingest/cell_metadata.py [hunk 1]
```diff
@@ -25,6 +25,7 @@ def __init__(self, file_path, file_id: str = None, study_accession: str = None):
         IngestFiles.__init__(self, file_path, self.ALLOWED_FILE_TYPES)
         self.headers = self.get_next_line(
             increase_line_count=False)
+        print(self.headers)
```
Comment (bistline) [inline ingest/cell_metadata.py]: Is there a reason we're printing these here?

File: ingest/clusters.py [hunk 2]
```diff
@@ -22,13 +22,13 @@ def __init__(self, file_path, *, name: str = None, study_accession: str = "", do
         self.source_file_type = 'cluster'
         self.has_z = 'z' in self.header
         self.top_level_doc = {
```
Comment (bistline) [inline ingest/clusters.py]: As we discussed, we need to add the `cell_annotations` dictionary to `top_level_doc`.  If it makes more sense to do this in a separate ticket/PR, that's fine so long as it gets added.

File: ingest/ingest_files.py [hunk 3]
```diff
@@ -10,32 +10,39 @@
...
 
 class IngestFiles:
-    def __init__(self, file_path, allowed_file_types, is_MTX=False):
+    def __init__(self, file_path, allowed_file_types, *, is_MTX=False, open_as=None):
         if not os.path.exists(file_path):
             raise IOError(f"File '{file_path}' not found")
         self.allowed_file_types = allowed_file_types
-        self.file_type, self.file = self.open_file(file_path)
+        self.file_type, self.file = self.open_file(file_path, open_as)
+        print(self.file_type)
```
Comment (bistline) [inline ingest/ingest_files.py]: Again, is there a reason for this print statement?

File: ingest/ingest_files.py [hunk 4]
```diff
@@ -106,6 +131,7 @@ def get_next_line(self, *, increase_line_count=True, split_line=True):
         """Returns a single line of txt, csv or tsv files"""
 
         next_row = next(self.file)
+        print(next_row)
```
Comment (bistline) [inline ingest/ingest_files.py]: Another print statement here that seems superfluous.

File: ingest/cell_metadata.py (hunk 1 above)
Comment (knapii-developments) [reply ingest/cell_metadata.py]: No, it was for debugging. Fixed in 6f04851

File: ingest/ingest_files.py (hunk 3 above)
Comment (knapii-developments) [reply ingest/ingest_files.py]: Nope. Addressed in 6f04851

File: ingest/ingest_files.py (hunk 4 above)
Comment (knapii-developments) [reply ingest/ingest_files.py]: It is. 6f04851

File: ingest/subsample.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        self.correct_annot_types()
+
+    def correct_annot_types(self):
+        """Ensures that numeric columns are rounded to 2 decimals points and
+        group annotations are strings
+        """
+        numeric_columns = self.file.xs("numeric", axis=1, level=1,
+                                       drop_level=False).columns.tolist()
+        self.file[numeric_columns] = self.file[numeric_columns].round(2)
+        group_columns = numeric_columns = self.file.xs("group", axis=1, level=1,
```
[the remark this answers is not in the document]
Comment (knapii-developments) [reply ingest/subsample.py]: It doesn't matter in this instance. But it's been fixed 6f04851

File: ingest/subsample.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                                     drop_level=False).columns.tolist()
+        self.file[group_columns] = self.file[group_columns].astype(str)
+
+    def prepare_cell_metadata(self):
+        """ Does an inner join on cell and cluster file """
+        if self.cell_metadata_file is not None:
+            self.merge_df(self.cell_metadata_file,
+                          self.file[self.coordinates_and_cell_names])
+
+    def dermine_coordinates_and_cell_names(self):
```
Comment (eweitz) [inline ingest/subsample.py]:
```suggestion
    def determine_coordinates_and_cell_names(self):
```

And similarly for other references to this method.

File: ingest/subsample.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            for index, df in enumerate(np.array_split(subset, 20)):
+                bin[str(index)] = df[self.coordinates_and_cell_names]
+        return bin, annotation
+
+    def subsample(self):
+        """Subsamples groups accross a given file"""
+        sample_sizes = [
+            sample_size for sample_size in self.SUBSAMPLE_THRESHOLDS if sample_size >= len(self.file.index)]
+        for bins in map(self.bin, self.columns):
+            # (name of current collumn)
```
Comment (eweitz) [inline ingest/subsample.py]:
```suggestion
            # (name of current column)
```

File: ingest/ingest_files.py [hunk 8]
```diff
@@ -10,32 +10,38 @@
...
         """ Opens txt, csv, or tsv formatted files"""
         open_file = open(file_path, encoding='utf-8-sig')
         file_connections = {
             # Remove BOM with encoding='utf-8-sig'
-            'text/csv': self.open_csv(open_file),
+            'text/csv': self.open_csv,
             'text/plain': open_file,
-            'text/tab-separated-values': self.open_tsv(open_file),
+            'text/tab-separated-values': self.open_tsv,
+            'pandas': self.open_pandas
```
Comment (eweitz) [inline ingest/ingest_files.py]: Suggestion until the future refactoring you mentioned:

```suggestion
            'dataframe': self.open_pandas
```

File: ingest/subsample.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+            annotation_name = bins[1]
+            # Looks like {"Unique value #1" : dataframe, "Unique value #2": dataframe,...}
+            anotation_dict = bins[0]
+            for sample_size in sample_sizes:
+                group_size = len(anotation_dict.keys())
+                # Dict of values for the x, y, and z coordinates
+                points = {k: [] for k in self.coordinates_and_cell_names}
+                num_per_group = int(sample_size / group_size)
```
[the remark this answers is not in the document]
Comment (knapii-developments) [reply ingest/subsample.py]: I believe I want it to round down. @bistline What are your opinions? How are we doing it now?

[the remark this answers is not in the document]
Comment (bistline) [reply ingest/subsample.py]: I don't think that it matters as any leftovers in the samples get added to the last group, which is by definition the largest sample population.  We're only talking about a difference of 1 entity anyway.

We're currently doing what Eno has (integer division).  `/` in Ruby discards the remainder and rounds down.

[the remark this answers is not in the document]
Comment (eweitz) [reply ingest/subsample.py]: That works!

File: ingest/ingest_pipeline.py [hunk 10]
```diff
@@ -14,14 +14,17 @@
 # Takes expression file and stores it into Firestore
 
 # Ingest cluster file
-python ingest_pipeline.py ingest_cluster --cluster-file ../tests/data/10k_cells_29k_genes.cluster.txt
+python ingest_pipeline.py ingest_cluster --cluster-file ../tests/data/10k_cells_29k_genes.cluster.txt --ingest-cluster
```
Comment (jlchang) [inline ingest/ingest_pipeline.py]: It's confusing to see "ingest_cluster" twice can you walk me through why this is needed?

File: ingest/ingest_pipeline.py [hunk 11]
```diff
@@ -14,14 +14,17 @@
...
 
 # Ingest Cell Metadata file
-python ingest_pipeline.py ingest_cell_metadata --cell-metadata-file ../tests/data/10k_cells_29k_genes.metadata.tsv
+python ingest_pipeline.py ingest_cell_metadata --cell-metadata-file ../tests/data/10k_cells_29k_genes.metadata.tsv --ingest-cell-metadata
```
Comment (jlchang) [inline ingest/ingest_pipeline.py]: The repetition of ingest-cell-metadata similarly confuses me...

File: ingest/ingest_pipeline.py [hunk 12]
```diff
@@ -251,6 +276,10 @@ def create_parser():
                                                   ' files are being ingested')
     parser_ingest_cluster.add_argument('--cluster-file',
                                        help='Path to cluster files')
+    parser_ingest_cluster.add_argument('--ingest-cluster-file', required=True,
+                                       choices=['True', 'False'],
```
[the remark this answers is not in the document]
Comment (knapii-developments) [reply ingest/ingest_pipeline.py]: I've changed to command to reflect this. Here's an example:  Check it [out](https://github.com/broadinstitute/scp-ingest-pipeline/blob/95703129e578d12b52ed121aa5328ac9edb438f6/ingest/ingest_pipeline.py#L275)!

**Pull request merged**